File: src/main/java/at/jku/ssw/coco/intellij/reference/CocoTokenReference.java
package at.jku.ssw.coco.intellij.reference;

import at.jku.ssw.coco.intellij.CocoUtil;
import at.jku.ssw.coco.intellij.psi.CocoTokenDecl;
import at.jku.ssw.coco.intellij.psi.HasCocoTokenReference;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class CocoTokenReference extends PsiReferenceBase<HasCocoTokenReference> implements PsiReference {

    public CocoTokenReference(@NotNull HasCocoTokenReference element, TextRange textRange) {
        super(element, textRange);
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        return CocoUtil.findTokenDecl(myElement.getContainingFile(), myElement.getTokenReferenceName());
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        List<CocoTokenDecl> characterDeclarations = CocoUtil.findTokenDecls(myElement.getContainingFile());
        return characterDeclarations.toArray();
    }
}

File: src/main/java/at/scheinecker/intellij/coco/reference/CocoTokenReference.kt
package at.scheinecker.intellij.coco.reference

import at.scheinecker.intellij.coco.CocoUtil
import at.scheinecker.intellij.coco.psi.HasCocoTokenReference
import com.intellij.codeInsight.lookup.LookupElementBuilder
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReference

class CocoTokenReference(element: HasCocoTokenReference, textRange: TextRange) : AbstractRenamableReference<HasCocoTokenReference>(element, textRange), PsiReference {

    override fun resolve(): PsiElement? {
        return CocoUtil.findTokenDecl(myElement.containingFile, myElement.name)
    }

    override fun getVariants(): Array<Any> {
        return CocoUtil.findTokenDecls(myElement.containingFile)
                .map { it ->
                    val tokenExpr = it.tokenExpr

                    val lookupElementBuilder = LookupElementBuilder.create(it)
                            .withTypeText("Token")

                    if (tokenExpr != null) {
                        return@map lookupElementBuilder.withTailText(" = " + tokenExpr.text, true)
                    }

                    lookupElementBuilder
                }
                .toTypedArray()
    }
}